aurora/contracts/AuroraErc20FastBridge.sol

```diff
@@ -0,0 +1,391 @@
+        TransferMessage memory transferMessage = _decodeTransferMessageFromBorsh(initTransferArgs);
+        require(
+            transferMessage.auroraSender == msg.sender,
+            "Aurora sender in transfer message doesn't equal to signer"
+        );
+
+        require(
+            _is_equal(transferMessage.transferTokenAddressOnNear, transferMessage.feeTokenAddressOnNear),
+            "The transfer and fee tokens are different. Different tokens not supported yet."
+        );
+
+        IEvmErc20 token = registeredTokens[transferMessage.transferTokenAddressOnNear];
+        require(address(token) != address(0), "The token is not registered!");
+        require(near.wNEAR.balanceOf(address(this)) > 0, "Not enough wNEAR balance of AuroraErc20FastBridge");
+
+        uint256 totalTokenAmount = uint256(transferMessage.transferTokenAmount + transferMessage.feeTokenAmount);
+
+        token.transferFrom(msg.sender, address(this), totalTokenAmount);
+
+        // WARNING: The `withdrawToNear` method works asynchronously.
+        // As a result, there is no guarantee that this method will be completed before `initTransfer`.
+        // In case of such an error, the user will be able to call `withdraw` method and get his/her tokens back.
+        // We expect such an error not to happen as long as transactions were executed in one shard.
+        token.withdrawToNear(bytes(getNearAddress()), totalTokenAmount);
+
+        string memory initArgsBase64 = Base64.encode(initTransferArgs);
+        bytes memory args = bytes(
+            string.concat(
+                '{"receiver_id": "',
+                bridgeAddressOnNear,
+                '", "amount": "',
+                Strings.toString(totalTokenAmount),
+                '", "msg": "',
+                initArgsBase64,
+                '"}'
+            )
+        );
+
+        PromiseCreateArgs memory callFtTransfer = near.call(
+            transferMessage.transferTokenAddressOnNear,
+            "ft_transfer_call",
+            args,
+            1,
+            INIT_TRANSFER_NEAR_GAS
+        );
+        bytes memory callbackArg = abi.encodeWithSelector(
+            this.initTokenTransferCallback.selector,
+            msg.sender,
+            initTransferArgs
+        );
+        PromiseCreateArgs memory callback = near.auroraCall(address(this), callbackArg, 0, BASE_NEAR_GAS);
+
+        callFtTransfer.then(callback).transact();
+    }
+
+    function initTokenTransferCallback(address signer, bytes memory initTransferArgs) public onlyRole(CALLBACK_ROLE) {
+        uint128 transferredAmount = 0;
+
+        if (AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful) {
+            transferredAmount = _stringToUint(AuroraSdk.promiseResult(0).output);
+        }
+
+        TransferMessage memory transferMessage = _decodeTransferMessageFromBorsh(initTransferArgs);
+
+        balance[transferMessage.transferTokenAddressOnNear][signer] += (transferMessage.transferTokenAmount +
```

Comment:
Lets add a check if the `(transferMessage.transferTokenAmount + transferMessage.feeTokenAmount -
transferredAmount)` to avoid extra write storage call